### PR TITLE
Fix bug when exception declared in filter is thrown

### LINKED ISSUE #4280: CA2200: Rethrow to preserve stack details - should not warn when throwing different exception
### Analyzer

**Diagnostic ID**: [CA2200](https://docs.microsoft.com/visualstudio/code-quality/ca2200): `Rethrow to preserve stack details`

### Analyzer source

**SDK**: [Built-in CA analyzers in .NET 5 SDK or later](https://docs.microsoft.com/dotnet/fundamentals/productivity/code-analysis)

**Version**: [SDK 5.0.100-preview.8](https://dotnet.microsoft.com/download/dotnet/5.0)

### Describe the bug
The caught exception `ex` is only thrown when it is mapped into a different exception by the `MapSendException(...)` method. The analyzer should not warn in this case, further I would assume CA2200 should not warn when the catch block is throwing a different exception (`mappedEx`) than the catched exception (`ex`)

```csharp
    try
    {
          ... 
    }
    catch (Exception ex) when (MapSendException(ex, cancellationToken, out Exception mappedEx))
    {
        throw mappedEx;
    }
```

The exact warning found [here](https://github.com/dotnet/runtime/blob/master/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs#L751), related to https://github.com/dotnet/runtime/pull/42957#discussion_r498532030

### Expected behavior
Should not warn in above case
### Actual behavior
Warning

### PR DESCRIPTION
**Pull request by Youssef1313**
Fixes #4280

### CODE REVIEW

File: src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/QualityGuidelines/RethrowToPreserveStackDetails.cs [hunk 1]
```diff
@@ -49,7 +49,8 @@ public override void Initialize(AnalysisContext analysisContext)
                     if (ancestor.Kind == OperationKind.CatchClause &&
                         ancestor is ICatchClauseOperation catchClause)
                     {
-                        if (catchClause.Locals.Contains(localReference.Local) &&
+                        if ((catchClause.ExceptionDeclarationOrExpression is not IVariableDeclaratorOperation variableDeclaratorOperation || variableDeclaratorOperation.Symbol == localReference.Local) &&
```
Comment (Evangelink) [inline src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/QualityGuidelines/RethrowToPreserveStackDetails.cs:52]: You should not be comparing the symbols with `==`.

**Review by Evangelink — APPROVED** [review]
Small suggestion for the symbol comparison but otherwise LGTM.

I totally didn't think about this scenario (never used it).

File: src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/QualityGuidelines/RethrowToPreserveStackDetails.cs (hunk 1 above)
Comment (Youssef1313) [inline src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/QualityGuidelines/RethrowToPreserveStackDetails.cs:52]:
```suggestion
                        if ((catchClause.ExceptionDeclarationOrExpression is not IVariableDeclaratorOperation variableDeclaratorOperation || SymbolEqualityComparer.Default.Equals(variableDeclaratorOperation.Symbol, localReference.Local)) &&
```

File: src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/QualityGuidelines/RethrowToPreserveStackDetails.cs [hunk 2]
```diff
@@ -49,7 +49,8 @@ public override void Initialize(AnalysisContext analysisContext)
                     if (ancestor.Kind == OperationKind.CatchClause &&
                         ancestor is ICatchClauseOperation catchClause)
                     {
-                        if (catchClause.Locals.Contains(localReference.Local) &&
+                        if ((catchClause.ExceptionDeclarationOrExpression is not IVariableDeclaratorOperation variableDeclaratorOperation || SymbolEqualityComparer.Default.Equals(variableDeclaratorOperation.Symbol, localReference.Local)) &&
```
Comment (mavasani) [inline src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/QualityGuidelines/RethrowToPreserveStackDetails.cs:52]: Why are we skipping the check if this not a variable declaration?

File: src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/QualityGuidelines/RethrowToPreserveStackDetailsTests.cs [hunk 3]
```diff
@@ -92,6 +92,30 @@ End Class
...
+using System;
+
+public abstract class C
+{
+    public void M()
+    {
+        try
+        {
+        }
+        catch (Exception ex) when (Map(ex, out Exception mappedEx))
```
Comment (mavasani) [inline src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/QualityGuidelines/RethrowToPreserveStackDetailsTests.cs:108]: Can you add a similar test where we have ‘catch(Exception) when M(out Exception ex)’ followed by ‘throw ex’? That should not fire.

File: src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/QualityGuidelines/RethrowToPreserveStackDetailsTests.cs [hunk 4]
```diff
@@ -92,6 +92,30 @@ End Class
...
+using System;
+
+public abstract class C
+{
+    public void M()
+    {
+        try
+        {
+        }
+        catch (Exception) when (Map(out Exception ex))
```
Comment (mavasani) [inline src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/QualityGuidelines/RethrowToPreserveStackDetailsTests.cs:108]: @Youssef1313 please add both tests - one in the repro and one that you added here based on PR feedback. Sorry if my comment was unclear.

Comment (Youssef1313) [reply src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/QualityGuidelines/RethrowToPreserveStackDetailsTests.cs:108]: @mavasani My bad. I thought I copied the existing test and modified the copy.

**Review by mavasani — APPROVED** [review]

**Pull request merged**